crates/store/src/filesystem/mod.rs

```diff
@@ -0,0 +1,412 @@
+                &encoded_meta,
+            )
+            .await
+            .unwrap();
+
+        self.fs
+            .root
+            .create_dir_all(&self.fs.value_root)
+            .map_err(Error::Io)
+            .unwrap();
+
+        let bin_link_path = format!(
+            "{}/{}/{}",
+            self.fs.key_root, self.key_identity, CONTENT_LINK_FILE
+        );
+        let bin_identity = utility::sha256(&self.buf);
+        let bin_value_path = format!("{}/{}.bin", self.fs.value_root, bin_identity);
+
+        if !self.fs.root.exists(&bin_value_path).await {
+            self.fs.write(&bin_value_path, &buf).await.unwrap();
+        }
+
+        self.fs
+            .root
+            .symlink(&format!("../../value/{}.bin", bin_identity), &bin_link_path)
+            .await
+            .unwrap();
+
+        (size, hash)
+    }
+}
+
+// TODO: abort and cleanup corrupted file structures properly: https://github.com/profianinc/drawbridge/issues/144
+#[async_trait]
+impl<K> Create<K> for Filesystem<K>
+where
+    K: Serialize + Sync + Send + Unpin + Eq + Hash,
+{
+    type Item<'a> = FilesystemCreateItem<'a, K> where K: 'a;
+    type Error = Error;
+
+    async fn create(
+        &mut self,
+        key: K,
+        mime: Mime,
+    ) -> Result<Self::Item<'_>, CreateError<Self::Error>> {
+        let (encoded_key, identity) = utility::encode_key(&key).map_err(CreateError::Internal)?;
+
+        if self.get_key_dir(&identity).await.is_ok() {
+            return Err(CreateError::Occupied);
+        }
+
+        self.root
+            .create_dir_all(format!("{}/{}", self.key_root, identity))
+            .map_err(Error::Io)
+            .map_err(CreateError::Internal)?;
+
+        self.write(
+            &format!("{}/{}/{}", self.key_root, identity, KEY_FILE),
+            &encoded_key,
+        )
+        .await
+        .map_err(CreateError::Internal)?;
+
+        Ok(FilesystemCreateItem {
+            fs: self,
+            key_identity: identity,
+            mime,
+            buf: vec![],
+        })
+    }
+}
+
+#[async_trait]
+impl<K> Get<K> for Filesystem<K>
+where
+    K: Serialize + Sync + Send + Eq + Hash,
+{
+    type Item<'a> = Cursor<Vec<u8>> where K: 'a;
+    type Error = Error;
+
+    async fn get(&self, key: K) -> Result<(Meta, Self::Item<'_>), GetError<Self::Error>> {
+        let (_encoded_key, identity) = utility::encode_key(&key).map_err(Error::into_get_error)?;
+        let key_dir = self
+            .get_key_dir(&identity)
+            .await
+            .map_err(Error::into_get_error)?;
+        let mut meta_file = key_dir.clone();
+        meta_file.push(META_FILE);
+        let mut bin_link_file = key_dir.clone();
+        bin_link_file.push(CONTENT_LINK_FILE);
+
+        let meta = self
+            .read_bincode(&meta_file)
+            .await
+            .map_err(|e| Error::BadMetaFile(identity.clone(), Box::new(e)).into_get_error())?;
+        let bin = self
+            .read_bytes(&bin_link_file)
```

Comment:
How will this work with chunked transfer?
Do we really want to read the whole file into memory on each request?

Reply:
We definitely want to avoid reading it all at once everytime to make it more robust but I think we should make this a follow up issue unless you know a quick way to get this working